sklearn/cluster/optics_.py

```diff
@@ -527,3 +555,379 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+        the region for an upward area.
+
+    mib : float
+        Maximum in between value, i.e. the maximum value between
+        the end of a steep down area and the current index. It is
+        irrelevant for steep up areas.
+    """
+    def __init__(self, start, end, maximum, mib):
+        self.start = start
+        self.end = end
+        # maximum is the beginning of a downward area
+        # and the end of an upward area
+        self.maximum = maximum
+        # maximum in between (ref Sec. 4.3.2)
+        self.mib = mib
+
+    def __repr__(self):
+        return ("start: %s, end: %s, max: %4g, mib: %4g" %
+                (self.start, self.end, self.maximum, self.mib))
+
+
+def _extend_region(steep_point, xward_point, start, min_samples):
+    """Extend the area until it's maximal.
+
+    It's the same function for both upward and downward reagions, depending on
+    the given input parameters. To extend an upward reagion,
+    ``steep_point=steep_upward`` and ``xward_point=downward`` are expected, and
+    to extend a downward region, ``steep_point=steep_downward`` and
+    ``xward_point=upward``.
+
+    Parameters
+    ----------
+    steep_point : bool array, shape (n_samples)
+        True if the point is an upward or downward steep point, respectively.
+
+    xward_point : bool array, shape (n_samples)
+        True if the point is upward or downward respectively.
+
+    start : integer
+        The start of the xward region.
+
+    min_samples : integer
+       The same as the min_samples given to OPTICS. Up and down steep
+       regions can't have more then ``min_samples`` consecutive non-steep
+       points.
+
+    Returns
+    -------
+    index : integer
+        The current index iterating over all the samples, i.e. where we are up
+        to in our search.
+
+    end : integer
+        The end of the region, which can be behind the index. The region
+        includes the ``end`` index.
+    """
+    n_samples = len(steep_point)
+    non_xward_points = 0
+    index = start
+    end = start
+    # find a maximal area
+    while index < n_samples:
+        if steep_point[index]:
+            non_xward_points = 0
+            end = index
+        elif not xward_point[index]:
+            # it's not a steep point, but still goes up.
+            non_xward_points += 1
+            # region should include no more than min_samples consecutive
+            # non steep xward points.
+            if non_xward_points > min_samples:
+                break
+        else:
+            return min(index, n_samples - 1), end
+
+        index += 1
+    return min(index + 1, n_samples - 1), end
+
+
+def _extend_upward_region(steep_upward_mask, downward_mask, start,
+                          min_samples):
+    return _extend_region(steep_upward_mask, downward_mask,
+                          start, min_samples)
+
+
+def _extend_downward_region(steep_downward_mask, upward_mask, start,
```

Comment:
This seems to be an exact copy of `_extend_upward_region`? Merge into `_extend_xward_region`?

Reply:
These two functions were added recently just to clarify the code, before there was only `_extend_xward_region`. These are just two aliases for clarity.

Reply:
The difference is the parameter naming

Reply:
Apparently these cause more confusion than they're worth